Yoast/wordpress-seo
# Identify form controls with error: add error messages and visual-semantic feedback

## Summary

This PR can be summarized in the following changelog entry:

* Add initial methods for server-side User Input Validation

## Relevant technical choices:
Builds on top of https://github.com/Yoast/wordpress-seo/pull/13258 to that should be merged first. Introduces a new `Yoast_Input_Validation` class with methods for user input validation and error reporting

## Test instructions
- switch to this branch 
- run `composer install`
- run `yarn && grunt build:css && grunt build:js && grunt webpack:buildProd`
Then, go to:
- SEO > General > Webmaster Tools
  - try, for example, with the following invalid values in this order: abc123.abc123 - abc123x - abc123.abc123 - abc123x
  - then try with valid values
- SEO > Social > Accounts
  - try the Twitter Username field, e.g.: enter an username with a space 
  - then try with valid values
  - to test the other URL fields, you should merge https://github.com/Yoast/wordpress-seo/pull/13240
- SEO > Social > Facebook 
  - try the Facebook App ID field: e.g. enter a value containing numbers and one or more letters
  - try a valid value, e.g. `246554168145`
- SEO > Social > Pinterest 
  - try the Pinterest confirmation field: e.g. enter a value containing a space or letters other than [a-f]
  - then try with valid values

In all cases:
If a pre-existing valid value was already saved, it gets restored and displayed again.
Same: if the field was empty, it gets restored to empty.
This is an existing behavior which hasn't changed in this PR. Currently under consideration by @thijs because from an UI perspective it doesn't seem to be an useful behavior, although ensuring a valid value is always stored in the DB is definitely valuable.

Things to check:
- the number of errors is correctly reported in the document `<title>`
- the input with error has an `aria-invalid="true"` attribute
- the input with error has an `aria-describedby` attribute referencing the ID of the paragraph with the error description
- the error description makes sense 
- the error description contains a final sentence with the invalid submitted value 
- the input field with error is correctly styled with colors accordingly to the design in https://github.com/Yoast/design/issues/323#issuecomment-501630532 

To run the tests:
There are now two new `@groups` and you can run just these test groups by issuing:
```
phpunit -c phpunit-integration.xml.dist --group input-validation
phpunit -c phpunit-integration.xml.dist --group option-social
```


Example screenshot: 

<img width="934" alt="Screenshot 2019-07-26 at 15 29 38" src="https://user-images.githubusercontent.com/1682452/61955010-6e405b00-afba-11e9-8c21-b1ea4cef1598.png">


## UI changes

## Documentation

## Quality assurance


Fixes https://github.com/Yoast/bugreports/issues/469
